exercises/robot-simulator/RobotSimulator.elm

```diff
@@ -1 +1,48 @@
-module RobotSimulator exposing (..)
+module RobotSimulator
+    exposing
+        ( Bearing(East, North, South, West)
+        , Robot
+        , advance
+        , defaultRobot
+        , simulate
+        , turnLeft
+        , turnRight
+        )
+
+
+type Bearing
+    = North
+    | East
+    | South
+    | West
+
+
+type alias Robot =
+    { bearing : Bearing
+    , coordinates : { x : Int, y : Int }
+    }
+
+
+defaultRobot : Robot
+defaultRobot =
+    Debug.crash "Please implement this function"
+
+
+turnRight : Robot -> Robot
+turnRight robot =
+    Debug.crash "Please implement this function"
+
+
+turnLeft : Robot -> Robot
+turnLeft robot =
+    Debug.crash "Please implement this function"
+
+
+advance : Robot -> Robot
+advance { bearing, coordinates } =
```

Comment:
@neganp when it comes to pattern matching, do you think keeping it like this is fine or should it be changed?

Reply:
Hmmm, good question. Would `advance ({ bearing, coordinates } as robot) =` be too confusing? It would certainly provide the most flexibility?

Reply:
I think the decision should be between `robot` and `{ bearing, coordinates }` instead of providing both. I think I'll actually just change it to `robot` because someone implementing this in the future will still be able to use `robot.bearing` or `robot.coordinates`.